source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+    // but if not we'll overwrite the deleted entry).
+    int probe_cap = INT_MAX;
+
+    int rv;
+    struct hash_table_entry *entry;
+    do {
+        int index = (hash_code + probe_idx) & state->mask;
+        entry = &state->slots[index];
+        if (!entry->hash_code) {
+            rv = AWS_ERROR_HASHTBL_ITEM_NOT_FOUND;
+            break;
+        }
+
+        if (entry->hash_code == hash_code && state->equals_fn(key, entry->element.key)) {
+            rv = AWS_ERROR_SUCCESS;
+            break;
+        }
+
+        int entry_probe = (index - entry->hash_code) & state->mask;
+
+        if (entry_probe < probe_idx) {
+            // We now know that our target entry cannot exist; if it did exist, it would be at the current location
+            // as it has a higher probe length than the entry we are examining and thus would have preempted that
+            // item
+            rv = AWS_ERROR_HASHTBL_ITEM_NOT_FOUND;
+            break;
+        }
+
+        probe_idx++;
+    } while(1);
+
+    *p_entry = entry;
+    if (p_probe_idx) {
+        if (probe_cap < probe_idx) {
+            probe_idx = probe_cap;
+        }
+        *p_probe_idx = probe_idx;
+    }
+
+    return rv;
+}
+
+int aws_common_hash_table_find(struct aws_common_hash_table *map,
+    const void *key, struct aws_common_hash_element **pElem) {
+    struct hash_table_state *state = map->pImpl;
+    uint64_t hash_code = hash_for(state, key);
+    struct hash_table_entry *entry;
+
+    int rv = find_entry(state, hash_code, key, &entry, NULL);
+
+    if (rv == AWS_ERROR_SUCCESS) {
+        *pElem = &entry->element;
+    }
+
+    return rv;
+}
+
+/*
+ * Attempts to find a home for the given entry. Returns after doing nothing if entry was not occupied.
+ */
+static struct hash_table_entry *emplace_item(struct hash_table_state *state, struct hash_table_entry entry, int probe_idx) {
+    struct hash_table_entry *initial_placement = NULL;
+
+    while (entry.hash_code) {
+        int index = (entry.hash_code + probe_idx) & state->mask;
+        struct hash_table_entry *victim = &state->slots[index];
+
+        int victim_probe_idx = (index - victim->hash_code) & state->mask;
+
+        if (!victim->hash_code || victim_probe_idx < probe_idx) {
+            if (!initial_placement) {
+                initial_placement = victim;
+            }
+
+            struct hash_table_entry tmp = *victim;
+            *victim = entry;
+            entry = tmp;
+
+            probe_idx = victim_probe_idx + 1;
+        } else {
+            probe_idx++;
+        }
+    }
+
+    return initial_placement;
+}
+
+static int expand_table(struct aws_common_hash_table *map) {
+    struct hash_table_state *old_state = map->pImpl;
+    struct hash_table_state template = *old_state;
+
+    update_template_size(&template, template.size * 2);
+
+    struct hash_table_state *new_state = alloc_state(&template);
+    if (!new_state) {
+        return AWS_ERROR_OOM;
```

Comment:
This could be cleaner and future granularity proof, if alloc_state called aws_raise_error and if this returns NULL then just return AWS_OP_ERR  and the error code will already be set.